redis/src/cluster_topology.rs

```diff
@@ -61,59 +62,98 @@ impl TopologyView {
 struct SlotMapValue {
     start: u16,
     addrs: SlotAddrs,
+    latest_used_replica: AtomicUsize,
 }
 
 impl SlotMapValue {
     fn from_slot(slot: Slot) -> Self {
         Self {
             start: slot.start(),
             addrs: SlotAddrs::from_slot(slot),
+            latest_used_replica: AtomicUsize::new(0),
         }
     }
 }
 
-#[derive(Debug, Default)]
-pub(crate) struct SlotMap(BTreeMap<u16, SlotMapValue>);
+#[derive(Debug, Default, Clone, PartialEq, Copy)]
+pub(crate) enum ReadFromReplicaStrategy {
+    #[default]
+    AlwaysFromPrimary,
+    RoundRobin,
+}
 
-impl SlotMap {
-    pub fn new() -> Self {
-        Self(BTreeMap::new())
-    }
+#[derive(Debug)]
+pub(crate) struct SlotMap {
+    slots: BTreeMap<u16, SlotMapValue>,
+    read_from_replica: ReadFromReplicaStrategy,
+}
 
-    pub fn fill_slots(&mut self, slots: Vec<Slot>) {
-        for slot in slots {
-            self.0.insert(slot.end(), SlotMapValue::from_slot(slot));
+fn get_address_from_slot(
+    slot: &SlotMapValue,
+    read_from_replica: ReadFromReplicaStrategy,
+    slot_addr: SlotAddr,
+) -> &str {
+    if slot_addr == SlotAddr::Master || slot.addrs.replicas.is_empty() {
+        return slot.addrs.primary.as_str();
+    }
+    match read_from_replica {
+        ReadFromReplicaStrategy::AlwaysFromPrimary => slot.addrs.primary.as_str(),
```

Comment:
maybe this should be documented. I agree it's not intuitive. 
`slot_addr` is where the command can go, `read_from_replica` represents whether the command can be sent to a replica. So, some commands can go to replicas, but only if `read_from_replica` is true.

Reply:
can't slot_addr be the address that is passed by command routing (as I wrote in the example above)?
We should allow sending non key based commands to replica if the user routes it this way

Reply:
I think that Route shouldn't be used for read from replica routing, but only for user routing / pre-fixed routing (e.g. allNodes). So get_address_from_slot would get the routing info and read_from_replica flag, and if routing isn't none - it will use the routing. Otherwise, it will get the address from the slot with read_from_replica flag that will need to decide if we should route it to the primary or to the replica.
Is it feasible?

Reply:
I'm sorry. I don't understand your suggesion. Can you write the signature you want `get_address_from_slot` to have?
and if `Route` is only for user routing, where/how will we get the slot for keyed commands, and whether they're readonly commands?